modules/apps/commerce/commerce-service/src/main/java/com/liferay/commerce/service/impl/CommerceOrderItemLocalServiceImpl.java

```diff
@@ -2653,6 +2766,169 @@ private void _validate(
 		}
 	}
 
+	private String _validateJSON(
+			CPDefinition cpDefinition, CPInstance cpInstance, String json)
+		throws PortalException {
+
+		JSONArray sanitizedJSONArray = _jsonFactory.createJSONArray();
+
+		if (cpDefinition != null) {
+			List<CPDefinitionOptionRel> cpDefinitionOptionRels =
+				_cpDefinitionOptionRelLocalService.getCPDefinitionOptionRels(
+					cpDefinition.getCPDefinitionId());
+			JSONObject jsonObject;
+			JSONArray originalJSONArray = CPJSONUtil.toJSONArray(json);
+
+			for (CPDefinitionOptionRel cpDefinitionOptionRel :
+					cpDefinitionOptionRels) {
+
+				jsonObject = _findMatchingJSONValue(
+					cpDefinitionOptionRel, originalJSONArray);
+
+				if ((cpDefinitionOptionRel.isRequired() ||
+					 cpDefinitionOptionRel.isSkuContributor()) &&
+					(jsonObject == null)) {
+
+					throw new CPDefinitionOptionRelException(
+						"Required option is missing");
+				}
+
+				if (jsonObject != null) {
+					_validateOriginalJSON(cpDefinitionOptionRel, jsonObject);
+				}
+
+				List<CPDefinitionOptionValueRel> cpDefinitionOptionValueRels =
+					_cpDefinitionOptionValueRelLocalService.
+						getCPDefinitionOptionValueRels(
+							cpDefinitionOptionRel.getCPDefinitionOptionRelId());
+
+				if (!cpDefinitionOptionValueRels.isEmpty()) {
+					for (CPDefinitionOptionValueRel cpDefinitionOptionValueRel :
+							cpDefinitionOptionValueRels) {
+
+						if (Validator.isNotNull(
+								cpDefinitionOptionValueRel.
+									getCPInstanceUuid())) {
+
+							jsonObject = _sanitizeJSON(
+								cpDefinitionOptionRel,
+								cpDefinitionOptionValueRel, cpInstance,
+								jsonObject);
+						}
+					}
+				}
+
+				if (jsonObject != null) {
+					sanitizedJSONArray.put(jsonObject);
+				}
+			}
+
+			if (cpDefinitionOptionRels.isEmpty() && !CPJSONUtil.isEmpty(json)) {
+				return originalJSONArray.toString();
```

Comment:
In some of our tests (e.g. testAddOrUpdateCommerceOrderItem()) we actually do not have any optionrels existing. This helps prevent us from completely losing any JSON provided, but I do recognize this means that the json is never validated. I'm not sure if there's a better way to handle that scenario or if we DO always expect an OptionRel to exist and I need to fix the test itself. Please let me know if you'd prefer a different outcome.